src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -256,93 +220,72 @@ private unsafe ValueTask WriteAsyncInternal(ReadOnlyMemory<byte> source, Cancell
+                Debug.Assert(vts._memoryHandle.Pointer != null);
 
-            // queue an async WriteFile operation and pass in a packed overlapped
-            int r = FileStreamHelpers.WriteFileNative(_fileHandle, source.Span, false, intOverlapped, out int errorCode);
-
-            // WriteFile, the OS version, will return 0 on failure.  But
-            // my WriteFileNative wrapper returns -1.  My wrapper will return
-            // the following:
-            // On error, r==-1.
-            // On async requests that are still pending, r==-1 w/ errorCode==ERROR_IO_PENDING
-            // On async requests that completed sequentially, r==0
-            // You will NEVER RELIABLY be able to get the number of bytes
-            // written back from this call when using overlapped IO!  You must
-            // not pass in a non-null lpNumBytesWritten to WriteFile when using
-            // overlapped structures!  This is ByDesign NT behavior.
-            if (r == -1)
-            {
-                // For pipes, when they are closed on the other side, they will come here.
-                if (errorCode == Interop.Errors.ERROR_NO_DATA)
+                long positionBefore = _filePosition;
+                if (CanSeek)
                 {
-                    // Not an error, but EOF. AsyncFSCallback will NOT be called.
-                    // Completing TCS and return cached task allowing the GC to collect TCS.
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-                    return ValueTask.CompletedTask;
+                    // Now set the position to read from in the NativeOverlapped struct
+                    // For pipes, we should leave the offset fields set to 0.
+                    nativeOverlapped->OffsetLow = (int)positionBefore;
+                    nativeOverlapped->OffsetHigh = (int)(positionBefore >> 32);
+
+                    // When using overlapped IO, the OS is not supposed to
+                    // touch the file pointer location at all.  We will adjust it
+                    // ourselves, but only in memory.  This isn't threadsafe.
+                    _filePosition += source.Length;
+                    UpdateLengthOnChangePosition();
                 }
-                else if (errorCode != Interop.Errors.ERROR_IO_PENDING)
-                {
-                    if (!_fileHandle.IsClosed && CanSeek)  // Update Position - It could be anywhere.
-                    {
-                        _filePosition = positionBefore;
-                    }
 
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-
-                    if (errorCode == Interop.Errors.ERROR_HANDLE_EOF)
+                // Queue an async WriteFile operation.
+                if (Interop.Kernel32.WriteFile(_fileHandle, (byte*)vts._memoryHandle.Pointer, source.Length, IntPtr.Zero, nativeOverlapped) == 0)
```

Comment:
same as `ReadFile`, it always returns `0` when async IO is enabled:

https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-writefile#return-value

Reply:
Ditto